File: codewave_py/closing_promp.py
import codewave_core.logger as logger
import codewave_core.util as util

class ClosingPromp():
	def __init__(self,codewave,start,end):
		self.codewave,self.start,self.end = codewave,start,end
		self.len = self.end - self.start
		self.found = False
		self.openBounds = self.closeBounds = None
	def begin(self):
		self.codewave.editor.beginUndoAction()
		self.codewave.editor.insertTextAt("\n"+self.codewave.brakets+self.codewave.closeChar+self.codewave.brakets,self.end)
		self.codewave.editor.insertTextAt(self.codewave.brakets+self.codewave.brakets+"\n",self.start)
		self.codewave.editor.endUndoAction()
		
		self.found = True
		p1 = self.start+len(self.codewave.brakets)
		p2 = self.end+len(self.codewave.brakets)*3+len(self.codewave.closeChar)+2
		self.openBounds = util.wrappedPos(  self.start, p1, p1, p1+len(self.codewave.brakets))
		self.closeBounds = util.wrappedPos( self.end, p2, p2, p2+len(self.codewave.brakets))
		if self.codewave.editor.allowMultiSelection() :
			self.codewave.editor.setMultiSel([util.Pos(p2,p2),util.Pos(p1,p1)])
			if self.codewave.editor.canListenToChange() :
				self.codewave.editor.addChangeListener(self.onChange)
		else : 
			self.codewave.editor.setCursorPos(p1)
		return self
	def onChange(self,ch = None):
		if ch is None or ch['ch'] ==  32:
			if self.shouldStop() :
				self.stop()
				self.cleanClose()
	def shouldStop(self):
		if not self.updateBounds() :
			return True
		else : 
			closeStr = self.codewave.editor.textSubstr(self.closeBounds.innerStart,self.closeBounds.innerEnd)
			if ' ' in closeStr :
				return True
		return False
	def cleanClose(self):
		if self.found :
			closeStr = self.codewave.editor.textSubstr(self.closeBounds.innerStart,self.closeBounds.innerEnd)
			self.codewave.editor.spliceText(self.closeBounds.innerStart,self.closeBounds.innerEnd,closeStr.strip())
			sels = self.codewave.editor.getMultiSel()
			change = False
			for sel in sels:
				if self.closeBounds.containsPos(sel) :
					sels.remove(sel)
					change = True
			if change :
				self.codewave.editor.setMultiSel(sels)
	def stop(self):
		if self.codewave.closingPromp == self :
			self.codewave.closingPromp = None
		self.codewave.editor.removeChangeListener(self.onChange)
	def cancel(self):
		if self.updateBounds() :
			self.codewave.editor.beginUndoAction()
			self.codewave.editor.spliceText(self.closeBounds.start-1,self.closeBounds.end,'')
			self.codewave.editor.spliceText(self.openBounds.start,self.openBounds.end+1,'')
			self.codewave.editor.endUndoAction()
		
			self.codewave.editor.setCursorPos(self.start,self.end)
		self.stop()
	def updateBounds(self):
		self.found = False
		self.openBounds = self.whithinOpenBounds(self.start+len(self.codewave.brakets))
		if self.openBounds is not None :
			self.closeBounds = self.whithinCloseBounds(self.openBounds)
			if self.closeBounds is not None :
				self.found = True
		return self.found;
	def whithinOpenBounds(self,pos):
		innerStart = self.start+len(self.codewave.brakets)
		if self.codewave.findPrevBraket(pos) == self.start and self.codewave.editor.textSubstr(self.start,innerStart) == self.codewave.brakets :
			innerEnd = self.codewave.findNextBraket(innerStart)
			if innerEnd is not None :
				return util.wrappedPos( self.start, innerStart, innerEnd, innerEnd+len(self.codewave.brakets))
	def whithinCloseBounds(self,openBounds):
		start = openBounds.end+self.len+2
		innerStart = start+len(self.codewave.brakets)+len(self.codewave.closeChar)
		if self.codewave.editor.textSubstr(start,innerStart) == self.codewave.brakets+self.codewave.closeChar :
			innerEnd = self.codewave.findNextBraket(innerStart)
			if innerEnd is not None :
				return util.wrappedPos( start, innerStart, innerEnd, innerEnd+len(self.codewave.brakets))

File: lib/ClosingPromp.js


const PosCollection = require("./positioning/PosCollection");

const Replacement = require("./positioning/Replacement");

const Pos = require("./positioning/Pos");

const OptionalPromise = require("./helpers/OptionalPromise");

var ClosingPromp = class ClosingPromp {
  constructor(codewave1, selections) {
    this.codewave = codewave1;
    this.timeout = null;
    this._typed = null;
    this.started = false;
    this.nbChanges = 0;
    this.selections = new PosCollection.PosCollection(selections);
  }

  begin() {
    this.started = true;
    return (0, OptionalPromise.optionalPromise)(this.addCarrets()).then(() => {
      if (this.codewave.editor.canListenToChange()) {
        this.proxyOnChange = (ch = null) => {
          return this.onChange(ch);
        };

        this.codewave.editor.addChangeListener(this.proxyOnChange);
      }

      return this;
    }).result();
  }

  addCarrets() {
    this.replacements = this.selections.wrap(this.codewave.brakets + this.codewave.carretChar + this.codewave.brakets + "\n", "\n" + this.codewave.brakets + this.codewave.closeChar + this.codewave.carretChar + this.codewave.brakets).map(function (p) {
      return p.carretToSel();
    });
    return this.codewave.editor.applyReplacements(this.replacements);
  }

  invalidTyped() {
    return this._typed = null;
  }

  onChange(ch = null) {
    this.invalidTyped();

    if (this.skipEvent(ch)) {
      return;
    }

    this.nbChanges++;

    if (this.shouldStop()) {
      this.stop();
      return this.cleanClose();
    } else {
      return this.resume();
    }
  }

  skipEvent(ch) {
    return ch != null && ch.charCodeAt(0) !== 32;
  }

  resume() {}

  shouldStop() {
    return this.typed() === false || this.typed().indexOf(' ') !== -1;
  }

  cleanClose() {
    var end, j, len, pos, repl, replacements, res, sel, selections, start;
    replacements = [];
    selections = this.getSelections();

    for (j = 0, len = selections.length; j < len; j++) {
      sel = selections[j];

      if (pos = this.whithinOpenBounds(sel)) {
        start = sel;
      } else if ((end = this.whithinCloseBounds(sel)) && start != null) {
        res = end.withEditor(this.codewave.editor).innerText().split(' ')[0];
        repl = new Replacement.Replacement(end.innerStart, end.innerEnd, res);
        repl.selections = [start];
        replacements.push(repl);
        start = null;
      }
    }

    return this.codewave.editor.applyReplacements(replacements);
  }

  getSelections() {
    return this.codewave.editor.getMultiSel();
  }

  stop() {
    this.started = false;

    if (this.timeout != null) {
      clearTimeout(this.timeout);
    }

    if (this.codewave.closingPromp === this) {
      this.codewave.closingPromp = null;
    }

    if (this.proxyOnChange != null) {
      return this.codewave.editor.removeChangeListener(this.proxyOnChange);
    }
  }

  cancel() {
    if (this.typed() !== false) {
      this.cancelSelections(this.getSelections());
    }

    return this.stop();
  }

  cancelSelections(selections) {
    var end, j, len, pos, replacements, sel, start;
    replacements = [];
    start = null;

    for (j = 0, len = selections.length; j < len; j++) {
      sel = selections[j];

      if (pos = this.whithinOpenBounds(sel)) {
        start = pos;
      } else if ((end = this.whithinCloseBounds(sel)) && start != null) {
        replacements.push(new Replacement.Replacement(start.start, end.end, this.codewave.editor.textSubstr(start.end + 1, end.start - 1)).selectContent());
        start = null;
      }
    }

    return this.codewave.editor.applyReplacements(replacements);
  }

  typed() {
    var cpos, innerEnd, innerStart;

    if (this._typed == null) {
      cpos = this.codewave.editor.getCursorPos();
      innerStart = this.replacements[0].start + this.codewave.brakets.length;

      if (this.codewave.findPrevBraket(cpos.start) === this.replacements[0].start && (innerEnd = this.codewave.findNextBraket(innerStart)) != null && innerEnd >= cpos.end) {
        this._typed = this.codewave.editor.textSubstr(innerStart, innerEnd);
      } else {
        this._typed = false;
      }
    }

    return this._typed;
  }

  whithinOpenBounds(pos) {
    var i, j, len, ref, repl, targetPos, targetText;
    ref = this.replacements;

    for (i = j = 0, len = ref.length; j < len; i = ++j) {
      repl = ref[i];
      targetPos = this.startPosAt(i);
      targetText = this.codewave.brakets + this.typed() + this.codewave.brakets;

      if (targetPos.innerContainsPos(pos) && targetPos.withEditor(this.codewave.editor).text() === targetText) {
        return targetPos;
      }
    }

    return false;
  }

  whithinCloseBounds(pos) {
    var i, j, len, ref, repl, targetPos, targetText;
    ref = this.replacements;

    for (i = j = 0, len = ref.length; j < len; i = ++j) {
      repl = ref[i];
      targetPos = this.endPosAt(i);
      targetText = this.codewave.brakets + this.codewave.closeChar + this.typed() + this.codewave.brakets;

      if (targetPos.innerContainsPos(pos) && targetPos.withEditor(this.codewave.editor).text() === targetText) {
        return targetPos;
      }
    }

    return false;
  }

  startPosAt(index) {
    return new Pos.Pos(this.replacements[index].selections[0].start + this.typed().length * (index * 2), this.replacements[index].selections[0].end + this.typed().length * (index * 2 + 1)).wrappedBy(this.codewave.brakets, this.codewave.brakets);
  }

  endPosAt(index) {
    return new Pos.Pos(this.replacements[index].selections[1].start + this.typed().length * (index * 2 + 1), this.replacements[index].selections[1].end + this.typed().length * (index * 2 + 2)).wrappedBy(this.codewave.brakets + this.codewave.closeChar, this.codewave.brakets);
  }

};
exports.ClosingPromp = ClosingPromp;
var SimulatedClosingPromp = class SimulatedClosingPromp extends ClosingPromp {
  resume() {
    return this.simulateType();
  }

  simulateType() {
    if (this.timeout != null) {
      clearTimeout(this.timeout);
    }

    return this.timeout = setTimeout(() => {
      var curClose, repl, targetText;
      this.invalidTyped();
      targetText = this.codewave.brakets + this.codewave.closeChar + this.typed() + this.codewave.brakets;
      curClose = this.whithinCloseBounds(this.replacements[0].selections[1].copy().applyOffset(this.typed().length));

      if (curClose) {
        repl = new Replacement.Replacement(curClose.start, curClose.end, targetText);

        if (repl.withEditor(this.codewave.editor).necessary()) {
          this.codewave.editor.applyReplacements([repl]);
        }
      } else {
        this.stop();
      }

      if (this.onTypeSimulated != null) {
        return this.onTypeSimulated();
      }
    }, 2);
  }

  skipEvent() {
    return false;
  }

  getSelections() {
    return [this.codewave.editor.getCursorPos(), this.replacements[0].selections[1] + this.typed().length];
  }

  whithinCloseBounds(pos) {
    var i, j, len, next, ref, repl, targetPos;
    ref = this.replacements;

    for (i = j = 0, len = ref.length; j < len; i = ++j) {
      repl = ref[i];
      targetPos = this.endPosAt(i);
      next = this.codewave.findNextBraket(targetPos.innerStart);

      if (next != null) {
        targetPos.moveSuffix(next);

        if (targetPos.innerContainsPos(pos)) {
          return targetPos;
        }
      }
    }

    return false;
  }

};
exports.SimulatedClosingPromp = SimulatedClosingPromp;

ClosingPromp.newFor = function (codewave, selections) {
  if (codewave.editor.allowMultiSelection()) {
    return new ClosingPromp(codewave, selections);
  } else {
    return new SimulatedClosingPromp(codewave, selections);
  }
};

